### PR TITLE
Allow query caching by default again

### LINKED ISSUE #33191: /_nodes/stats/indices/query_cache query_cache always empty
Elasticsearch version (bin/elasticsearch --version):6.4.0

Plugins installed: [analysis-ik]

JVM version (java -version):1.8.0_181

OS version (uname -a if on a Unix-like system):CentOS 7 4.18.3-1.el7.elrepo.x86_64

Description of the problem including expected versus actual behavior:

We have all cache metrics 0, such as:

/_nodes/stats/indices/query_cache

```
{
  "_nodes" : {
    "total" : 3,
    "successful" : 3,
    "failed" : 0
  },
  "cluster_name" : "weixinhost",
  "nodes" : {
    "6f0Kvy9VQf6pUdKpGc55_w" : {
      "timestamp" : 1535441875213,
      "name" : "node-2",
      "transport_address" : "172.31.100.174:9300",
      "host" : "172.31.100.174",
      "ip" : "172.31.100.174:9300",
      "roles" : [
        "master",
        "data",
        "ingest"
      ],
      "attributes" : {
        "ml.machine_memory" : "16026021888",
        "hdsize" : "300Gb",
        "xpack.installed" : "true",
        "az" : "cn-north-1a",
        "vmsize" : "r4.large",
        "ml.max_open_jobs" : "20",
        "ml.enabled" : "true"
      },
      "indices" : {
        "query_cache" : {
          "memory_size_in_bytes" : 0,
          "total_count" : 0,
          "hit_count" : 0,
          "miss_count" : 0,
          "cache_size" : 0,
          "cache_count" : 0,
          "evictions" : 0
        }
      }
    },
    ...
}
```
**Steps to reproduce**:

elasticsearch.yml

```
cluster.name:XXX
node.name: node-1
path.data: /elasticsearch/data
path.logs: /elasticsearch/log
network.host: 0.0.0.0
http.port: 9200
discovery.zen.ping.unicast.hosts: [.............]
discovery.zen.minimum_master_nodes: 2
bootstrap.memory_lock: true
```

/INDEX/_settings

```
{
  "INDEX" : {
    "settings" : {
      "index" : {
        "refresh_interval" : "10s",
        "number_of_shards" : "9",
        "provided_name" : "follower-v2",
        "creation_date" : "1535191750932",
        "requests" : {
          "cache" : {
            "enable" : "true"
          }
        },
        "unassigned" : {
          "node_left" : {
            "delayed_timeout" : "5m"
          }
        },
        "number_of_replicas" : "0",
        "queries" : {
          "cache" : {
            "enabled" : "true"
          }
        },
        "uuid" : "CeBrVgGRQ42NA4QlgQNiLA",
        "version" : {
          "created" : "6030299"
        }
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by jasontedor**
With the introduction of the default distribution, it means that by default the query cache is wrapped in the security implementation of the query cache. This cache does not allow caching if the request does not carry indices permissions. Yet, this will not happen if authorization is not allowed, which it is not by default. This means that with the introduction of the default distribution, query caching was disabled by default! This commit addresses this by checking if authorization is allowed and if not, delegating to the default indices query cache. Otherwise, we proceed as before with security. Additionally, we clear the cache on license state changes.

Closes #33191

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java [hunk 1]
```diff
@@ -50,6 +59,11 @@ public void clear(String reason) {
 
     @Override
     public Weight doCache(Weight weight, QueryCachingPolicy policy) {
+        if (licenseState.isAuthAllowed() == false) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: Is it intentional that this is looking only at `isAuthAllowed`  and not `isSecurityEnabled` ?

The implementation here will disable the cache for trial licenses even if they have not enabled security.

**Review by tvernum — COMMENTED** [review]
I left a question,

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java (hunk 1 above)
Comment (jaymode) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: This should check `isSecurityEnabled`. The other methods like isAuthAllowed do not take into account the default distribution behavior where security is disabled with a new trial license unless it is explicitly enabled.

**Review by jaymode — CHANGES_REQUESTED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java (hunk 1 above)
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: I do not think this is correct. I think that `LicenseState#isSecurityEnabled` will be true for a basic license.

Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: I think that it should be `licenseState.isSecurityEnabled() == false || licenseState.isAuthAllowed() == false`

Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: @jaymode @tvernum I pushed 7bae967a2edbd245f5e9db9ea0a8634388aa70c4.

**Review by jaymode — APPROVED** [review]
LGTM

**Comment by leeeboo** [comment]
@jasontedor Thanks!

**Pull request merged**